[iOS] Cannot scroll page by dragging on a bare <video> element

jernoble opened the pull request:
#### 960e0d109099517acdff834b0f7858b2c9de68d7
<pre>
[iOS] Cannot scroll page by dragging on a bare &lt;video&gt; element
<a href="https://bugs.webkit.org/show_bug.cgi?id=261563">https://bugs.webkit.org/show_bug.cgi?id=261563</a>
rdar://114720841

Reviewed by Simon Fraser.

The UI-side compositing subsystem relies on UIViews of particular classes or conforming to
particular protocols for collecting the appropriate views during hit testing. WebAVPlayerLayerView
is not one of those clasess, nor can it conform to the protocols defined in WebKit (as it&apos;s a
product of WebCore). So to the hit testing logic of UI-side compositing, it&apos;s a black hole.

Add a new class, WKVideoView, which is a WKCompositingView as well as WKNativelyInteractible, to
host the WebAVPlayerLayerView.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
(WebKit::VideoFullscreenManagerProxy::createViewWithID):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeHostIOS.mm:
* Source/WebKit/UIProcess/ios/WKVideoView.h: Added.
* Source/WebKit/UIProcess/ios/WKVideoView.mm: Added.
(+[WKVideoView layerClass]):
(-[WKVideoView hitTest:withEvent:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/268041@main">https://commits.webkit.org/268041@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/923c730314a84d9a42fcea7cfefe0adc914d8016

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18435 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18774 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19369 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20285 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17262 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22081 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18925 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/19158 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18660 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18844 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16056 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21164 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16067 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16813 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23290 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17087 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16983 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21177 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17562 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14901 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16643 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/16646 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4394 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21009 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17401 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
```diff
@@ -730,15 +732,21 @@ - (BOOL)prefersStatusBarHidden
         if (![[view layer] superlayer])
             [playerLayer addSublayer:[view layer]];
 
+        RetainPtr videoView = [[WKVideoView alloc] initWithFrame:CGRectMake(0, 0, initialSize.width(), initialSize.height())];
```
```suggestion
        auto videoView = adoptNS([[WKVideoView alloc] initWithFrame:CGRectMake(0, 0, initialSize.width(), initialSize.height())]);
```

Without `adoptNS`, this is a leak.

smfr commented on Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
```diff
@@ -730,15 +732,21 @@ - (BOOL)prefersStatusBarHidden
         if (![[view layer] superlayer])
             [playerLayer addSublayer:[view layer]];
 
+        RetainPtr videoView = [[WKVideoView alloc] initWithFrame:CGRectMake(0, 0, initialSize.width(), initialSize.height())];
```
Leak. You need adoptRef()

jernoble replied:
Oof, good catch.

smfr commented on Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
```diff
@@ -730,15 +732,21 @@ - (BOOL)prefersStatusBarHidden
 
+        RetainPtr videoView = [[WKVideoView alloc] initWithFrame:CGRectMake(0, 0, initialSize.width(), initialSize.height())];
+        [videoView addSubview:playerView.get()];
+        [NSLayoutConstraint activateConstraints:@[
```
Makes me a bit nervous.

jernoble replied:
I can change this to just have the WKVideoView resize its subviews to have frames equal to its bounds.

smfr commented on Source/WebKit/UIProcess/ios/WKVideoView.mm:
```diff
@@ -0,0 +1,41 @@
+
+- (UIView *)hitTest:(CGPoint)point withEvent:(UIEvent *)event
+{
+    return [self pointInside:point withEvent:event] ? self : nil;
```
Is it true that we never need to hit-test descendants of this view?

jernoble replied:
It's true. The only contents of this view will be AVPlayerLayerView, and its only contents are a layer-hosting view. All should have bounds equal to their superview's frame.

smfr reviewed (approved):
Would be nice to have an API test.

rr-codes commented on Source/WebKit/UIProcess/ios/WKVideoView.mm:
```diff
@@ -0,0 +1,41 @@
+
+#if PLATFORM(IOS_FAMILY)
+@implementation WKVideoView
++ (Class)layerClass
```
Why is this overridden method needed?

jernoble replied:
To avoid creating a WKCompositingLayer as the backing of this view; it's not needed as the contents will never be painted.

jernoble committed:
[iOS] Cannot scroll page by dragging on a bare <video> element
https://bugs.webkit.org/show_bug.cgi?id=261563
rdar://114720841

Reviewed by Simon Fraser.

The UI-side compositing subsystem relies on UIViews of particular classes or conforming to
particular protocols for collecting the appropriate views during hit testing. WebAVPlayerLayerView
is not one of those clasess, nor can it conform to the protocols defined in WebKit (as it's a
product of WebCore). So to the hit testing logic of UI-side compositing, it's a black hole.

Add a new class, WKVideoView, which is a WKCompositingView as well as WKNativelyInteractible, to
host the WebAVPlayerLayerView.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/VideoFullscreenManagerProxy.mm:
(WebKit::VideoFullscreenManagerProxy::createViewWithID):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeHostIOS.mm:
* Source/WebKit/UIProcess/ios/WKVideoView.h: Added.
* Source/WebKit/UIProcess/ios/WKVideoView.mm: Added.
(+[WKVideoView layerClass]):
(-[WKVideoView hitTest:withEvent:]):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/268041@main
